### PR TITLE
Including transactions in expense response

### PR DESCRIPTION
**Pull request by djavorek**
## Description of Changes
- Added expense response DTO
- Created new transaction operation to query by expense id (no user authorization there)
- Calling new transaction operation from controller

> ❗ PR base is `25-server-side-testing`

### Issue
- #46 

### Response of /api/expense/1

```
[
    {
        "id": 1,
        "name": "Netflix for my little family",
        "description": "Cheapest monthly plan of Netflix - Maybe upgrade later",
        "value": "24.90",
        "start_date": "2022-09-24",
        "user_id": 1,
        "currency_type_id": 1,
        "recurrence_type_id": 1,
        "predefined_expense_id": 1,
        "transactions": [
            {
                "id": 1,
                "donor_name": "Kate",
                "value": "0.05",
                "date": "2022-09-29",
                "currency_type_id": 2,
                "expense_id": 1
            },
            {
                "id": 2,
                "donor_name": "David",
                "value": "7.00",
                "date": "2022-10-23",
                "currency_type_id": 2,
                "expense_id": 1
            }
        ]
    },
    {
        "id": 2,
        "name": "Synology C2 backup",
        "description": "Its not much but it keeps our photos safe",
        "value": "34.99",
        "start_date": "2022-03-15",
        "user_id": 1,
        "currency_type_id": 2,
        "recurrence_type_id": 2,
        "predefined_expense_id": null,
        "transactions": [
            {
                "id": 3,
                "donor_name": "Wifey",
                "value": "50.00",
                "date": "2022-04-12",
                "currency_type_id": 2,
                "expense_id": 2
            },
            {
                "id": 4,
                "donor_name": "My colleague who use exotic Hungarian Forint",
                "value": "1000.00",
                "date": "2022-05-13",
                "currency_type_id": 1,
                "expense_id": 2
            }
        ]
    }
]
```

### CODE REVIEW

File: server/api/src/dtos/expenses.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    pub start_date: Date,
+    pub user_id: Id,
+    pub currency_type_id: Id,
+    pub recurrence_type_id: Id,
+    pub predefined_expense_id: Option<Id>,
+    pub transactions: Vec<transaction::Model>,
+}
+
+impl ExpenseWithTransactions {
+    fn from(&self, expense: expense::Model, transactions: Vec<transaction::Model>) -> self {
```
Comment (szattila98) [inline server/api/src/dtos/expenses.rs:21]: This from function somehow feels odd, I will search for a way to make it more idiomatic, good enough for now tough.

File: server/core/src/dto/expenses.rs [hunk 2]
```diff
@@ -47,3 +47,36 @@ pub struct NewPredefinedExpenseRequest {
     pub currency_type_id: Id,
     pub recurrence_type_id: Id,
 }
+
+pub type ExpenseResponse = Vec<ExpenseWithTransactions>;
```
Comment (szattila98) [inline server/core/src/dto/expenses.rs:51]: Aliases are usually named after the function which returns them, so this one should be `GetExpensesResponse`.

File: server/api/src/routes/expenses.rs [hunk 3]
```diff
@@ -1,23 +1,39 @@
...
     State(ref conn): State<DatabaseConnection>,
     Path(user_id): Path<Id>,
-) -> Result<Json<Vec<expense::Model>>, ErrorMsg<()>> {
+) -> Result<Json<ExpenseResponse>, ErrorMsg<()>> {
     match expense_operations::find_expenses_by_user_id(conn, user.id, user_id).await {
-        Ok(expenses) => Ok(Json(expenses)),
+        Ok(expenses) => {
+            let mut expense_response: ExpenseResponse = vec![];
+
+            for expense in expenses {
```
Comment (szattila98) [inline server/api/src/routes/expenses.rs:24]: I am not exactly sure I like DTO construction in the "controller" layer but maaaaybe it is good? I genuinely don't know.

File: server/core/src/dto/expenses.rs [hunk 4]
```diff
@@ -47,3 +47,36 @@ pub struct NewPredefinedExpenseRequest {
...
+
+#[derive(Clone, Serialize)]
+pub struct ExpenseWithTransactions {
```
Comment (szattila98) [inline server/core/src/dto/expenses.rs:54]: Why is there a struct named ExpenseWithTransactions both in api and core? It's a bit strange design wise.

File: server/core/src/logic/transaction_operations.rs [hunk 5]
```diff
@@ -44,6 +44,18 @@ pub async fn create_transaction(
     Ok(transaction.id)
 }
 
+pub async fn find_transactions_by_expense_id(
+    conn: &DatabaseConnection,
+    expense_id: Id,
+) -> Result<Vec<transaction::Model>, DbErr> {
+    let transactions = Transaction::find()
```
Comment (szattila98) [inline server/core/src/logic/transaction_operations.rs:51]: Maybe the `find_related` function would be better for this. More [here](https://www.sea-ql.org/SeaORM/docs/basic-crud/select/#find-related-models).

**Review by szattila98 — CHANGES_REQUESTED** [review]

File: server/api/src/dtos/expenses.rs (hunk 1 above)
Comment (djavorek) [reply server/api/src/dtos/expenses.rs:21]: This leftover file is getting deleted.

File: server/core/src/dto/expenses.rs (hunk 2 above)
Comment (djavorek) [reply server/core/src/dto/expenses.rs:51]: Deleting this file now.

File: server/core/src/dto/expenses.rs (hunk 4 above)
Comment (djavorek) [reply server/core/src/dto/expenses.rs:54]: Yes, the real one is in the core.

File: server/api/src/routes/expenses.rs (hunk 3 above)
Comment (djavorek) [reply server/api/src/routes/expenses.rs]: Refactored it into logic.

File: server/core/src/logic/transaction_operations.rs (hunk 5 above)
Comment (djavorek) [reply server/core/src/logic/transaction_operations.rs]: Hey Mister, using `find_with_related` now. Also modified your test according to my changes.
See my comment for the test below.

File: server/core/src/logic/expense_operations.rs [hunk 6]
```diff
@@ -1,34 +1,44 @@
...
-    let expenses = Expense::find()
+    let mut result: Vec<ExpenseWithTransactions> = vec![];
+    let expenses_with_transactions = Expense::find()
+        .find_with_related(Transaction)
         .filter(expense::Column::UserId.eq(user_id))
         .all(conn)
         .await?;
-    Ok(expenses)
+
+    for expense in expenses_with_transactions {
```
Comment (szattila98) [inline server/core/src/logic/expense_operations.rs:37]: I would like to see a more functional approach here. Also I think you can destructure expense like so `for (expense, transactions) in expenses_with_transactions ...` and the tuple numbering would be more readable.

Comment (djavorek) [reply server/core/src/logic/expense_operations.rs]: Fixed

**Review by szattila98 — APPROVED** [review]
LGTM, nice changes, thank you for your work. 🚀

**Pull request merged**